Create new Subscription in pkg/apis/eventing as per the new spec.

vaikas-google opened the pull request:
Fixes #

## Proposed Changes

Add a new version of Subscription CRD as per the new Spec.

**** To reduce code churn, tests are incomplete Please only look at the subscription_types for now****

**Release Note**
<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->
```release-note
```

scothis commented:
👍 on the general direction.

One idea worth considering. Instead of using strings for these properties we could use ObjectReferences. The corresponding resolved values for `Processor` and `To` can be added to SubscriptionStatus, so that dispatchers still don't need to resolve in the data plane. A centralized subscription controller can take responsibility for synchronizing the resolutions. Webhook validation can reject invalid configurations.

Advantages:
- higher level resources, like Flow, are easier to implement because they don't need to preform the resolution themselves
- resolutions are automatically resynchronized over time. More important for non-controlled subscriptions
- any component that needs to resolve these targets will need rbac get access to the target component. Centralizing the resolution also enables centralizing the privilege. Adding a new CRD means adding a single new Role/RoleBinding.
- easy to add support for common resources that are unlikely to implement our targetable interface (like k8s core Service)
- future: a ClusterChannel CRD was discussed as a structured way to pass events between namespaces. An ObjectReference makes it easier to support other types in the future. This should be very rare though.
- future: access controls could be applied to the subscription

Disadvantages:
- ObjectReferences are more complex (forces all creators to know the resource kind)
- Subscription creators would be unable to create custom resolution strategies, unless we introduce a manual mode
- there is a delay between when a target resource changes its `domainInternal` status and when the subscription is re-synced. The current Flow resource also suffers this delay, however, provisioners with foreknowledge are less able to trigger an immediate update.

n3wscott commented on pkg/apis/channels/v1alpha1/subscription_types.go:
```diff
@@ -57,14 +63,17 @@ type SubscriptionSpec struct {
+	// Processor is the processor service DNS name. Events
+	// from the From channel will be delivered here and replies are sent
+	// to To channel.
+	Processor string `json:"processor,omitempty"`
```
How about `Call`?

markfisher replied:
since kubernetes resource yaml describes the desired _declarative_ state to be reconciled by a controller, I personally prefer to avoid using **verbs** that suggest the yaml is defining the _imperative_ logic

n3wscott replied:
`calls` ?

evankanderson replied:
From conversation in Slack, I'd like to see names which provide the following:
1. Clearly unambiguous about the sequence (`from`, `processor`, `to` seems ambiguous whether `to` or `processor` is last).
2. Work cleanly for all three cases (`from`, `processor`, `to`), (`from`, `processor`), (`from`, `to`).

Mostly, we seem to think `from` is clear. Here are some options for the other two:
* `processor`: `call`, `invoke`, `through`, `subscriber`, `apply`
* `to`: `finally` (collides with exception handling), `then` (may not be strong enough), `result`, `out`, `continue` (may collide with loops and not imply last item).

markfisher replied:
Here's an option where there's always a to/from pair at top-level, if both `processor` and `channel` are included in the `to`, the `processor` will always be invoked prior to sending its result to the channel.

```
from: x
to:
  processor: foo
  channel: y

from: x
to:
  processor: foo

from: x
to:
  channel: y
```

The downside is that it adds a level of nesting in the config, but that might be worth the tradeoff since these resources would typically be generated from higher level resources and/or CLIs.

vaikas-google replied:
What's the upside of adding this extra level of nesting, since to: block will still have the same behaviour as if we kept it the top level?

vaikas-google commented:
@scothis making them references is appealing for sure. I was left wondering how to actually then configure the underlying 'real channel' to wire things together. It's one thing to resolve the reference, but then once the wiring happens, how do you see that happening. Seems like once you resolve the referenced object, you'd need to modify that object to actually create the subscription. So, if a BetterChannel CRD was created and subscription resolved that it was the underlying channel, how would the SubscriptionController then know how to modify the BetterChannel resource?

scothis commented:
@vaikas-google It may help to talk about the `processor` and `to` fields separately from the `from` field for this discussion. The `processor` and `to` fields are resolving from an ObjectRef to a domain name via the Targetable interface.

The `from` field is a bit more open ended since the Provisioner for the Channel is also the Provisioner for Subscription for that Channel. As an initial pass, I think we should force the ObjectRef Kind to `channels.eventing.knativedev/v1alpha1` rejecting any other values. The Channel Provisioner is responsible for selecting which Subscriptions are for a Channel it provisioned and then provision the Subscription. For the stub bus this is by indexing the Subscription for the Channel in the dispatcher. For the Pub/Sub bus this is by creating a Subscription for the Topic in Cloud Pub/Sub and then receiving messages for that Subscription.

I don't think the SubscriptionController would ever modify the Channel resource, it would only be responsible for updating the SubcriptionStatus for the values it resolves. The Channel Provisioner is responsible for provisioning/reconciling the Subscription with the resolved properties.

vaikas-google commented:
But the SubscriptionController needs to somehow communicate to the resolved channel that there's a new subscription?
I think it's odd that the ChannelController would be looking at a spec for an object it does not control.
I'd expect something like this (only From considered here):
  1. Subscription is created: From is resolved to FromChannel
  2. SubscriptionController modifies FromChannel.Spec to add this new subscription
  3. ChannelController sees the FromChannel.Spec and does the thing it needs to do

Ideally (I think) Channel would have a subresource that the SubscriptionController would twiddle, but in lieue of that, we could make it so that the "contract" (given that we do refs, and we need to have some commonality there in the future) would be that a target of a ChannelRef resolve to an object that in it's spec has a Subscriptions List (or something like that), and the SubscriptionController then can just Read Modify Write the list by adding (or removing on deletion) the new (or deleted) subscriber.

And yes, rejecting anything other than channels.eventing.knative.dev/v1alpha1 for now is fine, just thinking how this could look like for superduperchannel that comes next and what that contract may look like.

Lastly, as per the document discussed, this will move to:
eventing.knative.dev/v1alpha1
I just made this WIP here to show only the minimal set of changes, once we agree on things, I'll move this to pkg/apis/eventing/v1alpha1/subscription_types

evankanderson commented on pkg/apis/channels/v1alpha1/subscription_validation_test.go:
```diff
@@ -30,34 +30,48 @@ func TestSubscriptionSpecValidation(t *testing.T) {
+		name: "missing to",
+		c: &SubscriptionSpec{
+			From: "fromChannel",
+			To:   "toChannel",
```
Should this be "processor", or should "to" be left off? (This matches the next test case right now.)

vaikas-google commented:
to->follower?

scothis commented on pkg/apis/channels/v1alpha1/subscription_types.go:
```diff
@@ -63,17 +63,48 @@ type SubscriptionSpec struct {
+	// to these events for further processing.
+	//
+	// For example, this could be a reference to a Route resource
+	// or a Configuration resource.
```
s/Configuration/Service/

Configuration is not targetable.

vaikas-google replied:
done

scothis commented on pkg/apis/channels/v1alpha1/subscription_types.go:
```diff
@@ -63,17 +63,48 @@ type SubscriptionSpec struct {
+	//   - Name
+	From *corev1.ObjectReference `json:"from,omitempty"`
 
 	// Processor is the processor service DNS name. Events
```
no longer a DNS name

vaikas-google replied:
done

vaikas-google commented:
@scothis yes, validation needs to change and things need to move to right location. Just trying to minimize churn and only focus on the API definition and semantics for now. Once we agree on that, I'll move this to eventing/pkg/apis/eventing/v1alpha1/ and do all the remaining work there.

scothis commented on pkg/apis/channels/v1alpha1/subscription_validation.go:
```diff
@@ -20,21 +20,29 @@ import (
+		return fe
+	}
+
+	if equality.Semantic.DeepEqual(ss.Processor, &corev1.ObjectReference{}) && equality.Semantic.DeepEqual(ss.To, &corev1.ObjectReference{}) {
```
could also cause issues if one is nil and the other is empty

scothis commented on pkg/apis/channels/v1alpha1/subscription_validation.go:
```diff
@@ -49,7 +57,7 @@ func (current *Subscription) CheckImmutableFields(og apis.Immutable) *apis.Field
 	}
 
-	ignoreArguments := cmpopts.IgnoreFields(SubscriptionSpec{}, "Subscriber", "Arguments")
+	ignoreArguments := cmpopts.IgnoreFields(SubscriptionSpec{}, "Processor", "Arguments")
```
`To` should also allow mutation, only `From` should be immutable. It was an oversight that `ReplyTo` is immutable right now.

scothis commented on pkg/apis/channels/v1alpha1/subscription_validation_test.go:
```diff
@@ -20,8 +20,49 @@ import (
+	ProcessorName     = "processor"
+)
+
+func getValidFromRef() *corev1.ObjectReference {
```
should also test invalid `From` and `To` ObjectReferences:
- non channel `Kind` and `APIVersion`
- fields other than `Kind`, `APIVersion` and `Name`

vaikas-google replied:
Working on it... Just want to resolve the names...

evankanderson commented on pkg/apis/channels/v1alpha1/subscription_types.go:
```diff
@@ -57,14 +64,53 @@ type SubscriptionSpec struct {
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
```
What about UID?

I'm assuming that Namespace is specifically excluded?

scothis replied:
> What about UID?

Is there a good way to lookup a resource by UID?

> I'm assuming that Namespace is specifically excluded?

Keeping subscriptions within a single namespace is a good idea until we understand fully understand the tenancy consequences of allowing them to cross namespaces.

vaikas-google replied:
Let's start with names (no UIDs), and a single namespace until we have more understanding on what cases this does not handel?

evankanderson commented on pkg/apis/channels/v1alpha1/subscription_validation.go:
```diff
@@ -20,21 +20,29 @@ import (
-		fe.Details = "the Subscription must reference a Channel"
+	if ss.From == nil || equality.Semantic.DeepEqual(ss.From, &corev1.ObjectReference{}) {
+		fe := apis.ErrMissingField("from")
+		fe.Details = "the Subscription must reference a from channel"
```
This should probably also ensure that the `ss.From.Name` and `ss.From.Kind` (at least) are non-empty.

(In general, it feels like there should be a function to validate an `ObjectReference` in the context of Subscription.)

vaikas-google replied:
le sigh :) Yes, there's bunch of more work to be done here :) Thanks for the comments, I just don't want to keep churning these things until we settle on the shape and the names. I'm going to add validation for now that it can only be a Channel for now as discussed in Slack yesterday, and later on we can relax.

evankanderson commented on pkg/apis/channels/v1alpha1/subscription_types.go:
```diff
@@ -57,14 +64,53 @@ type SubscriptionSpec struct {
+	//   - APIVersion
+	//   - Name
+	// +optional
+	Processor *corev1.ObjectReference `json:"processor,omitempty"`
```
Is the idea that you would need to create a k8s Service to be able to call something which is not on the k8s cluster (e.g. a VM)? Previously we had the idea that a targetable was either a DNS name or an ObjectReference.

vaikas-google replied:
Yeah, so, this is a bit interesting in a sense, that I'd expect that if you're hitting something that's outside the cluster, just a DNS entry is probably not going to cut it. I'd expect things like credentials to be involved on that, so as not to just be wide open to everything. I could hoist this into a higher level object to give us some future proofing for now.

scothis replied:
Another option is to allow the reference for a targetable to also include a k8s Service. It doesn't implement the targetable interface, but we know how to shoehorn it into what we need.

vaikas-google replied:
@scothis when we talked about this wrt Targetable, thinking was that a service was not enough, that it would actually have to be a URI. So, I'm hoisting that up.

evankanderson commented on pkg/apis/channels/v1alpha1/subscription_types.go:
```diff
@@ -57,14 +64,53 @@ type SubscriptionSpec struct {
+	//   - APIVersion
+	//   - Name
+	// +optional
+	To *corev1.ObjectReference `json:"to,omitempty"`
```
I can't comment on other parts of this CRD, so leaving the comment here:

1. Can we remove `Arguments` until we need them, and prefer to have Subscription-specific arguments rather than Channel-specific `Arguments`? (See also `s/[]Arguments/runtime.RawExtension/`.)
2. What conditions do we need? I think we probably want at least 2 Types: `Ready` and `ReferencesResolved`, where [`Ready` is the canonical "happy state" type](https://github.com/knative/serving/blob/master/docs/spec/errors.md#knative-serving-condition-type).
   * Do we need `LastUpdateTime` and `LastTransitionTime`? I'd be inclined to leave off that bookkeeping until we have evidence we need it.

scothis replied:
> Can we remove `Arguments` until we need them

Yes, we haven't come across a use case for argument/params with the current buses.

vaikas-google replied:
removed.

scothis commented:
A **subscription** triggers sending events from a **channel** to a **subscriber**; the subscriber's **replies** (if any) are forwarded to a **channel** for subsequent processing.

```yaml
spec:
  channel: <channel-ref>
  subscriber: <targetable-ref>
  replyStrategy:
    channel: <channel-ref>
```

The key elements of the SubscriptionSpec are the channel being subscribed to and the subscriber being targeted. Handling replies from the subscriber are managed by a strategy for which we have an two implementations:
- forwarding to a hardwired channel
- "/dev/null" where the reply is logged and tossed

As discussed in Slack we want the ability to define other strategies for mapping or dynamic resolution.

vaikas-google commented:
@scothis this was the idea that you'd have Subscription coordinate with the Function what the return headers are and then dispatchers will dynamically route to different channels based on that?

scothis commented:
> @scothis this was the idea that you'd have Subscription coordinate with the Function what the return headers are and then dispatchers will dynamically route to different channels based on that?

That was the shape of the discussion in Slack. The exact details still need to firm up, but this SubscriptionSpec proposal leaves room for other strategies to be plugged in.

For example, a hypothetical router could look like:

```yaml
spec:
  channel: <channel-ref>
  subscriber: <targetable-ref>
  replyStrategy:
    mapping:
    - name: widget
      channel: <channel-ref>
    - name: gadget
      channel: <channel-ref>
```

vaikas-google commented:
Wheee :)
Ok, I think I addressed the comments relating to the CRD shape from @evankanderson and @scothis 
Please only look at the Spec and the structs defined within it. I don't want to keep churning the other files until we resolve the shape.

evankanderson commented:
@scothis Another option for dynamic dispatch would be to use sticks + spokes to configure things as follows:

```
[channel] --> [colorizer] --> [channel] --> [only-blue] -->
                                        \-> [only-green] -->
```

I'd be inclined to leave `to` as a simple field for now, and provide reply strategies as a separate subscription field once we have a better idea what the spectrum of useful options is. (For example, I think `only-blue` and `only-green` are useful in a pre-call `filter` step, while some others prefer to treat them as functions.)

grantr commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,200 @@
+
+const (
+	// Dispatching means the subscription is actively listening for incoming events on its channel and dispatching them.
+	SubscriptionDispatching SubscriptionConditionType = "Dispatching"
```
Is the meaning of `Dispatching` different from the meaning of `Ready` in serving?

vaikas-google replied:
I haven't touched the status, it's old cruft. Wanted to get Spec right first.

grantr commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,200 @@
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	// This field is immutable
```
Why is this field immutable?

scothis replied:
The idea was that since a Subscription cannot be unsubscribed from one channel and subscribed to another channel atomically that we shouldn't pretend to offer atomic mutations. There would be a (hopefully brief) period where either both channels have the subscription or neither channel has the subscription. Preventing mutation forces the user to think about the semantics they want during the migration.

scothis commented:
As I understand the current state of this PR, a fairly typical Subscription will look like:

```yaml
apiVersion: eventing.knative.dev/v1alpha1
kind: Subscription
metadata:
  name: <name>
  namespace: <namespace>
spec:
  from:
    apiVersion: eventing.knative.dev/v1alpha1
    kind: Channel
    name: <channel-name>
  # at least one of:
  call:
    # one of:
    target:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: <service-name>
    # or
    targetURI: <service-uri>
  # or
  result:
    target:
      apiVersion: eventing.knative.dev/v1alpha1
      kind: Channel
      name: <channel-name>
status:
  conditions:
  - lastTransitionTime: 2018-08-28T22:54:51Z
    lastUpdateTime: 2018-08-28T22:54:51Z
    message: <message>
    reason: Synced
    status: "True"
    type: Dispatching
```

compared to the current Subscription:

```yaml
apiVersion: channels.knative.dev/v1alpha1
kind: Subscription
metadata:
  name: <name>
  namespace: <namespace>
spec:
  channel: <channel-name>
  subscriber: <service-dns-name>
  replyTo: <channel-dns-name> # optional
status:
  conditions:
  - lastTransitionTime: 2018-08-28T22:54:51Z
    lastUpdateTime: 2018-08-28T22:54:51Z
    message: <message>
    reason: Synced
    status: "True"
    type: Dispatching
```

I think there are good, justified reasons for the extra verbosity, but it is a good bit more complex.

n3wscott commented:
/lgtm

> but it is a good bit more complex.

This gives us an out for Channel-like things to be switched in-place of channels... I think it is worth that future flexibility and I don't think we are expecting users to make subscriptions directly (they are a result of the higher order objects).

markfisher commented:
I'd like to make one last push for preferring *nouns* since the resource is describing desired state in a declarative way. I had proposed "channel"->"subscriber" as the essence of what a Subscription describes, and the "replyStrategy" would describe how to handle a reply from the subscriber.

That said, if we are going to stick with "call" then I would prefer consistency for what is currently "result" (i.e. if one is a verb, they both should be). The latter could be "forward" or "reply" (I would probably choose "reply" since that is commonly used for describing this role in messaging systems - e.g. via "replyTo" headers - even if it's not immediately replying back to an original caller but instead continuing to flow through a sequence of subsequent channels/subscribers).

scothis commented:
> This gives us an out for Channel-like things to be switched in-place of channels.

I want to be clear that I do not think we should support "channel-like things". The thinking was that we may have other kinds of channels in the future. The only concrete thought I have in this direction is a ClusterChannel to enable eventing between namespaces in a clearly defined way. Same behavior, but a separate CRD because that's how k8s works.

grantr commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,200 @@
+	"github.com/knative/pkg/webhook"
+	"k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Non-blocking nit: can this be renamed to `metav1` for consistency?

vaikas-google commented:
Ok, I think this is ready for a review. My most sincere apologies if I missed some comments in the flurry of discussions, here, doc and slack.

vaikas-google commented on pkg/apis/eventing/v1alpha1/subscription_defaults.go:
```diff
@@ -0,0 +1,25 @@
+}
+
+func (ss *SubscriptionSpec) SetDefaults() {
+	// TODO anything?
```
For consistency with other files, I'm leaving this in for now. If you feel strongly, I can remove.

vaikas-google commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// Callable specifies the reference to an object that's expected to
+// provide the resolved target of the action. Currently we inspect
```
done

vaikas-google commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,204 @@
+// This ensures that we can support external targets and for ease of use
+// we also allow for an URI to be specified.
+// There of course is also a requirement for the resolved Callable to
+// behave properly at the data plane level.
```
done. I felt like we should have a pointer to a spec and have a link to it here. We don't, so added what you proposed.

vaikas-google commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,204 @@
+// we can utilize a dynamic client instead of having to understand all
+// kinds of different types of objects. As long as they adhere to this
+// particular contract, they can be used as a Target.
+// This ensures that we can support external targets and for ease of use
```
done

vaikas-google commented on pkg/apis/eventing/v1alpha1/subscription_types.go:
```diff
@@ -0,0 +1,204 @@
+type SubscriptionConditionType string
+
+const (
+	// SubscriptionReady is when the From,Channel and Result have been reconciled successfully.
```
done.

markfisher replied:
@evankanderson sure "call" *can* be used as a noun, but IMO "call" is going to be interpreted as a verb by anyone who uses this, especially when combined with "from", which is a preposition and the logical starting point among the components involved thereby setting an imperative context (also as I mentioned earlier, when creating a subscription you are actually subscribing TO a channel, not from).

I think all 3 should be nouns, and not ambiguous ones. I'm personally not even convinced the current "channel" and "subscriber" should change, since those are the most straightforwardly descriptive names for the 2 primary roles. Everything related to the handling of a subscriber's response is a side-effect, except in the (rare) "bridge" case, and for that I think it's best to view as a no-op subscriber since a channel should not be interpreted as the direct subscriber of another channel (even if the "bridge" is logical with no physical resource, it's enforcing the tinker-toy conceptual model). I do see how "replyChannel" or something similar would be a better alternative to "replyTo" since the latter presents the same problem of an imperative context, and if we are going to support dynamic channel resolution in addition to hardwired literal values, then "replyStrategy" could be the top-level as @scothis proposed earlier (with the "mapping" example). 

Sorry if this seems like bike-shedding. I do not view it that way, since these are not names used in an internal implementation. When defining the resources that users interact with, the names of those resources and their properties are among the most important things to get right.

vaikas-google commented on pkg/apis/eventing/v1alpha1/subscription_validation.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+func isValidCallable(c Callable) *apis.FieldError {
+	if c.TargetURI != nil && c.Target != nil && !equality.Semantic.DeepEqual(c.Target, &corev1.ObjectReference{}) {
```
Something in my readings led me to believe you needed to do this. I could totally be wrong about this however.

vaikas-google commented:
I'd like to unblock this work and suggest that we take the naming
conversation to another thread. We still have tons of work to do before we
can have a working end-to-end implementation and to unblock controller
development, and provisioner implementation and so forth, I'd like to
suggest that if the shape is right, we'll check this in and  continue
naming discussion and have an issue to revisit names before the release.
Sound good?

On Fri, Sep 14, 2018 at 6:05 AM Mark Fisher <notifications@github.com>
wrote:

> *@markfisher* commented on this pull request.
> ------------------------------
>
> In pkg/apis/eventing/v1alpha1/subscription_types.go
> <https://github.com/knative/eventing/pull/421#discussion_r217705324>:
>
> > +	// This object must fulfill the Sinkable contract.
> +	//
> +	// TODO: Specify the required fields the target object must
> +	// have in the status.
> +	// You can specify only the following fields of the ObjectReference:
> +	//   - Kind
> +	//   - APIVersion
> +	//   - Name
> +	// +optional
> +	Target *corev1.ObjectReference `json:"target,omitempty"`
> +}
> +
> +type SubscriptionConditionType string
> +
> +const (
> +	// SubscriptionReady is when the From,Channel and Result have been reconciled successfully.
>
> @evankanderson <https://github.com/evankanderson> sure "call" *can* be
> used as a noun, but IMO "call" is going to be interpreted as a verb by
> anyone who uses this, especially when combined with "from", which is a
> preposition and the logical starting point among the components involved
> thereby setting an imperative context (also as I mentioned earlier, when
> creating a subscription you are actually subscribing TO a channel, not
> from).
>
> I think all 3 should be nouns, and not ambiguous ones. I'm personally not
> even convinced the current "channel" and "subscriber" should change, since
> those are the most straightforwardly descriptive names for the 2 primary
> roles. Everything related to the handling of a subscriber's response is a
> side-effect, except in the (rare) "bridge" case, and for that I think it's
> best to view as a no-op subscriber since a channel should not be
> interpreted as the direct subscriber of another channel (even if the
> "bridge" is logical with no physical resource, it's enforcing the
> tinker-toy conceptual model). I do see how "replyChannel" or something
> similar would be a better alternative to "replyTo" since the latter
> presents the same problem of an imperative context, and if we are going to
> support dynamic channel resolution in addition to hardwired literal values,
> then "replyStrategy" could be the top-level as @scothis
> <https://github.com/scothis> proposed earlier (with the "mapping"
> example).
>
> Sorry if this seems like bike-shedding. I do not view it that way, since
> these are not names used in an internal implementation. When defining the
> resources that users interact with, the names of those resources and their
> properties are among the most important things to get right.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/knative/eventing/pull/421#discussion_r217705324>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKwedMPjlgNXqmUORoWJjV8Z1xrefFDNks5ua6mRgaJpZM4WfoH6>
> .
>

markfisher commented:
/lgtm
